errai-navigation/src/main/java/org/jboss/errai/ui/nav/client/local/Navigation.java

```diff
@@ -444,49 +448,71 @@ public void onAttachOrDetach(final AttachEvent event) {
 
+        accessEvent.fireAsync(component, new LifecycleCallback() {
           @Override
           public void callback(final boolean success) {
             if (success) {
-              locked = true;
+              NavigationControl showControl = new NavigationControl(Navigation.this, new Runnable() {
+                @Override
+                public void run() {
+                  try {
+                    // Fire IOC lifecycle event to indicate that the state of the
+                    // bean has changed.
+                    // TODO make this smarter and only fire state change event when
+                    // fields actually changed.
+                    stateChangeEvent.fireAsync(component);
+
+                    setCurrentPage(request.pageNode);
+                    currentWidget = componentWidget;
+                    currentComponent = component;
+                    navigatingContainer.setWidget(componentWidget);
+                    request.pageNode.pageShown(component, request.state);
+                  } finally {
+                    locked = false;
+                  }
+
+                  handleQueuedRequests(request, fireEvent);
+                }
+              }, new Runnable() {
+                @Override
+                public void run() {
+                  locked = false;
+                }
+              });
+
               try {
+                locked = true;
                 hideCurrentPage();
-                request.pageNode.pageShowing(component, request.state);
-
-                // Fire IOC lifecycle event to indicate that the state of the
-                // bean has changed.
-                // TODO make this smarter and only fire state change event when
-                // fields actually changed.
-                stateChangeEvent.fireAsync(component);
-
-                setCurrentPage(request.pageNode);
-                currentWidget = componentWidget;
-                currentComponent = component;
-                navigatingContainer.setWidget(componentWidget);
-                request.pageNode.pageShown(component, request.state);
-              } finally {
+                request.pageNode.pageShowing(component, request.state, showControl);
+              }
+              catch (Exception ex) {
                 locked = false;
+                throw ex;
               }
-
-              handleQueuedRequests(request, fireEvent);
-            }
-            else {
+            } else {
               request.pageNode.destroy(component);
             }
           }
         });
       }
     });
 
-    if (currentPage != null && currentWidget != null && currentComponent != null && currentWidget.asWidget() == navigatingContainer.getWidget()) {
-      currentPage.pageHiding(Factory.maybeUnwrapProxy(currentComponent), control);
-    }
-    else {
+    if (!hiding && currentPage != null && currentWidget != null && currentComponent != null && currentWidget.asWidget() == navigatingContainer.getWidget()) {
```

Comment:
The purpose is when calling `redirect` in `@PageHiding` this causes infinite recursion if we don't ensure the hiding phase isn't executed multiple times. The `hiding` var will always be set to `false` when `proceed` is invoked so there shouldn't be able issues.

As for your use case you are right that would be a problem, any suggestions on the best way to handle this with the use case presented above too?